dev/src/index.ts

```diff
@@ -1198,15 +1219,100 @@ export class Firestore implements firestore.Firestore {
     this.bulkWritersCount -= 1;
   }
 
+  /**
+   * Recursively deletes all documents and subcollections at and under the
+   * specified level.
+   *
+   * If any deletes fail, the promise is rejected with an error message
+   * containing the number of failed writes and the stack trace of the last
+   * failed write. The provided reference is deleted regardless of whether
+   * all deletes succeeded, except when Firestore fails to fetch the provided
+   * reference's descendants.
+   *
+   * Firestore uses a BulkWriter instance with default settings to perform the
+   * deletes. To customize throttling rates or add success/error callbacks,
+   * pass in a custom BulkWriter instance.
+   *
+   * @param ref The reference of a document or collection to delete.
+   * @param bulkWriter Custom BulkWriter instance with which to perform the
+   * deletes with.
+   * @return A promise that resolves when all deletes have been performed.
+   * The promise is rejected if any of the deletes fail.
+   */
+  recursiveDelete<T = DocumentData>(
+    ref: CollectionReference<T> | DocumentReference<T>,
+    bulkWriter?: BulkWriter
+  ): Promise<void> {
+    const docStream = this.getAllDescendants(ref);
+    const writer = bulkWriter ?? this.getBulkWriter();
+    const deleteCompleted = new Deferred<void>();
+    let errorCount = 0;
+    let lastError: Error | undefined;
+
+    // Capture the error stack to preserve stack tracing across async calls.
+    const stack = Error().stack!;
+
+    docStream
+      .on('error', err => {
+        err.code = Status.FAILED_PRECONDITION;
+        err.message =
+          'Failed to fetch children documents. ' +
+          'The provided reference was not deleted.';
+        deleteCompleted.reject(wrapError(err, stack));
+      })
+      .on('data', (snap: QueryDocumentSnapshot) => {
+        const docRef = new DocumentReference(this, snap.ref._path);
+        writer.delete(docRef).catch(err => {
+          errorCount++;
+          lastError = err;
+        });
+      })
+      .on('end', () => {
+        writer.flush().then(async () => {
+          if (ref instanceof DocumentReference) {
+            try {
+              await ref.delete();
```

Comment:
I can't think of a more elegant way to differentiate between the final document reference delete failing vs. the BulkWriter deletes. Any suggestions?

Reply:
I haven't looked at the rest of the PR, but if you want to work on it before I do - couldn't you just let the BulkWriter handle this delete as well?

Reply:
I still have a few more changes/cleanup I need to push upstream, so reviewing now might result in some wasted work.

Thanks for the suggestion! I originally figured a normal delete is less verbose than enqueuing another operation to BulkWriter and calling flush(), but using BulkWriter allows for any user callbacks to also run.